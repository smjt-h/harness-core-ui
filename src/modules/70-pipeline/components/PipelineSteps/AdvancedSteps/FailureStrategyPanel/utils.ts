/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { SetStateAction } from 'react'
import type { FormikErrors, FormikProps } from 'formik'
import { isEmpty } from 'lodash-es'
import { Intent } from '@blueprintjs/core'

import type {
  RetryFailureActionConfig,
  IgnoreFailureActionConfig,
  ManualInterventionFailureActionConfig,
  AbortFailureActionConfig,
  StepGroupFailureActionConfig,
  MarkAsSuccessFailureActionConfig,
  StageRollbackFailureActionConfig,
  FailureStrategyConfig,
  OnFailureConfig
} from 'services/cd-ng'

export type AllActions =
  | RetryFailureActionConfig
  | IgnoreFailureActionConfig
  | ManualInterventionFailureActionConfig
  | AbortFailureActionConfig
  | StepGroupFailureActionConfig
  | MarkAsSuccessFailureActionConfig
  | StageRollbackFailureActionConfig

export interface AllFailureStrategyConfig extends FailureStrategyConfig {
  onFailure: OnFailureConfig & { action: AllActions }
}

export function hasItems<T>(data?: T[]): boolean {
  return Array.isArray(data) && data.length > 0
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function findTabWithErrors(errors?: FormikErrors<any>): number {
  const fsErrors = errors?.failureStrategies

  return Array.isArray(fsErrors)
    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
      fsErrors.findIndex((err: any) => !isEmpty(err))
    : -1
}

export interface FormState {
  failureStrategies?: AllFailureStrategyConfig[]
}

export interface HandleChangeInStrategiesProps {
  formValues: FormState
  selectedStrategyNum: number
  setSelectedStrategyNum: (value: SetStateAction<number>) => void
  setFormikState: FormikProps<FormState>['setFormikState']
}

export function handleChangeInStrategies({
  formValues,
  selectedStrategyNum,
  setSelectedStrategyNum,
  setFormikState
}: HandleChangeInStrategiesProps): void {
  const fs = formValues.failureStrategies
  /* istanbul ignore else */
  if (Array.isArray(fs)) {
    /* istanbul ignore else */
    if (selectedStrategyNum >= fs.length) {
      // select the new last tab, if the last tab was deleted
      setSelectedStrategyNum(Math.max(0, fs.length - 1))
    }

    /* istanbul ignore else */
    if (fs.length === 0) {
      // reset errors when all the tabs are deleted
      setFormikState({ errors: {}, submitCount: 0 })
    }
  }
}

export function getTabIntent(i: number, selectedStrategyNum: number): Intent {
  return i === selectedStrategyNum ? Intent.PRIMARY : Intent.NONE
}
