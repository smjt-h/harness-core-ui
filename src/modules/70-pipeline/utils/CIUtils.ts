/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { get, isEmpty } from 'lodash-es'
import moment from 'moment'
import { RUNTIME_INPUT_VALUE, SelectOption } from '@harness/uicore'
import type { GitFilterScope } from '@common/components/GitFilters/GitFilters'
import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { RegExAllowedInputExpression } from '@pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet'
import type { GitQueryParams } from '@common/interfaces/RouteInterfaces'
import { useQueryParams } from '@common/hooks'
import type { StepViewType } from '@pipeline/components/AbstractSteps/Step'
import { ConnectorRefWidth } from './constants'

// Returns first 7 letters of commit ID
export function getShortCommitId(commitId: string): string {
  return commitId.slice(0, 7)
}

// TODO: Add singular forms, better using i18n because they have support for it
export function getTimeAgo(timeStamp: number): string {
  const currentDate = moment(new Date())
  const timeStampAsDate = moment(timeStamp)

  if (currentDate.diff(timeStampAsDate, 'days') > 30) {
    return `on ${timeStampAsDate.format('MMM D')}`
  } else if (currentDate.diff(timeStampAsDate, 'days') === 1) {
    return 'yesterday'
  } else if (currentDate.diff(timeStampAsDate, 'days') === 0) {
    if (currentDate.diff(timeStampAsDate, 'minutes') >= 60) {
      return `${currentDate.diff(timeStampAsDate, 'hours')} hours ago`
    } else {
      return `${currentDate.diff(timeStampAsDate, 'minutes')} minutes ago`
    }
  } else {
    return `${currentDate.diff(timeStampAsDate, 'days')} days ago`
  }
}

export function useGitScope(): GitFilterScope | undefined {
  const gitDetails = usePipelineContext()?.state?.gitDetails
  const { repoIdentifier, branch } = useQueryParams<GitQueryParams>()

  if (!isEmpty(gitDetails)) {
    return {
      repo: gitDetails.repoIdentifier!,
      branch: gitDetails.branch!,
      getDefaultFromOtherRepo: true
    }
  } else if (!!repoIdentifier && !!branch) {
    return {
      repo: repoIdentifier,
      branch,
      getDefaultFromOtherRepo: true
    }
  }
}

export const getAllowedValuesFromTemplate = (template: Record<string, any>, fieldPath: string): SelectOption[] => {
  if (!template || !fieldPath) {
    return []
  }
  const value = get(template, fieldPath, '')
  const items: SelectOption[] = []
  if (RegExAllowedInputExpression.test(value as string)) {
    // This separates out "<+input>.allowedValues(a, b, c)" to ["<+input>", ["a", "b", "c"]]
    const match = (value as string).match(RegExAllowedInputExpression)
    if (match && match?.length > 1) {
      const allowedValues = match[1]
      items.push(...allowedValues.split(',').map(item => ({ label: item, value: item })))
    }
  }
  return items
}

export const shouldRenderRunTimeInputView = (value: any): boolean => {
  if (!value) {
    return false
  }
  if (typeof value === 'object') {
    return Object.keys(value).some(key => typeof value[key] === 'string' && value[key].startsWith(RUNTIME_INPUT_VALUE))
  } else {
    return typeof value === 'string' && value.startsWith(RUNTIME_INPUT_VALUE)
  }
}

export const shouldRenderRunTimeInputViewWithAllowedValues = (
  fieldPath: string,
  template?: Record<string, any>
): boolean => {
  if (!template || !fieldPath) {
    return false
  }
  const allowedValues = get(template, fieldPath, '')
  return shouldRenderRunTimeInputView(allowedValues) && RegExAllowedInputExpression.test(allowedValues)
}

export const getConnectorRefWidth = (viewType: StepViewType): number =>
  Object.entries(ConnectorRefWidth).find(key => key[0] === viewType)?.[1] || ConnectorRefWidth.DefaultView
