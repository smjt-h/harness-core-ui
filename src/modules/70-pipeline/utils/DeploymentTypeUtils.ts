/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import type { ServiceDefinition } from 'services/cd-ng'
import type { StringKeys } from 'framework/strings'

export const deploymentTypeLabel: Record<ServiceDefinition['type'], StringKeys> = {
  Kubernetes: 'kubernetesText',
  NativeHelm: 'pipeline.nativeHelm',
  Ssh: 'SSH',
  ServerlessAwsLambda: 'pipeline.serviceDeploymentTypes.serverlessAwsLambda',
  WinRm: 'pipeline.serviceDeploymentTypes.winrm'
}

export const deploymentTypeIcon: Record<string, IconName> = {
  Kubernetes: 'service-kubernetes',
  NativeHelm: 'service-helm',
  ServerlessAwsLambda: 'service-serverless-aws',
  Ssh: 'secret-ssh',
  WinRm: 'command-winrm'
}
