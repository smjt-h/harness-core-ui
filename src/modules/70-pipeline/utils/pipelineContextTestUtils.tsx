/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

/**
 * !IMPORTANT
 * Point to Note: Go to the previous versions to see how the functions are actually implemented.
 * Add them back only if they are being unit tested, else leave them out.
 * This is because it unecessarily branches out coverage and we cannot test those that are implemented if they aren't used
 */
import React from 'react'
import { Route } from 'react-router-dom'

import { pick } from 'lodash-es'

import { GitSyncTestWrapper, GitSyncTestWrapperProps } from '@common/utils/gitSyncTestUtils'

import {
  PipelineContext,
  PipelineContextInterface
} from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'

import pipelineContextProviderProps from './__tests__/mockJson/pipelineContextProvider.json'

export interface PipelineContextWrapperProps extends GitSyncTestWrapperProps {
  pipelineContextValues?: Partial<PipelineContextInterface>
}

jest.mock('services/portal', () => ({
  useGetDelegateSelectorsUpTheHierarchy: jest.fn
}))

jest.mock('services/cd-ng', () => ({
  useGetConnectorListV2: jest.fn(),
  useGetConnectorList: jest.fn(),
  useGetConnector: jest.fn(),
  useGetTestConnectionResult: jest.fn(),
  useGetTestGitRepoConnectionResult: jest.fn(),
  useGetSteps: jest.fn(),
  useGetExecutionStrategyList: jest.fn(),
  useGetExecutionStrategyYaml: jest.fn(),
  useGetServiceList: jest.fn(),
  useGetEnvironmentListForProject: jest.fn(),
  useListGitSync: jest.fn(),
  useGetStepsV2: jest.fn()
}))

jest.mock('services/pipeline-ng', () => ({
  useGetYamlSchema: jest.fn(),
  getPipelinePromise: jest.fn(),
  useGetSteps: jest.fn(),
  useCreateVariablesV2: jest.fn()
}))

export function PipelineContextTestWrapper(
  props: React.PropsWithChildren<PipelineContextWrapperProps>
): React.ReactElement {
  const { defaultAppStoreValues, pipelineContextValues } = props
  return (
    <GitSyncTestWrapper
      {...pick(props, ['path', 'pathParams', 'queryParams'])}
      defaultAppStoreValues={{
        featureFlags: {},
        selectedProject: {
          identifier: 'dummy',
          name: 'dummy',
          modules: ['CD']
        },
        isGitSyncEnabled: true,
        connectivityMode: 'DELEGATE',
        ...defaultAppStoreValues
      }}
    >
      <PipelineContext.Provider
        value={
          {
            fetchPipeline: jest.fn,
            getStageFromPipeline: jest.fn,
            getStagePathFromPipeline: jest.fn,
            setSelection: jest.fn,
            setTemplateTypes: jest.fn,
            updatePipeline: jest.fn,
            updatePipelineView: jest.fn,
            addNewNodeToMap: jest.fn,
            setStepsGraphCanvasState: jest.fn,
            setSelectedStepId: jest.fn,
            setSelectedStageId: jest.fn,
            renderPipelineStage: jest.fn,
            ...pipelineContextProviderProps,
            ...pipelineContextValues,
            state: {
              ...pipelineContextProviderProps.state,
              ...pipelineContextValues?.state
            }
          } as PipelineContextInterface
        }
      >
        <Route exact path={props.path}>
          {props.children}
        </Route>
      </PipelineContext.Provider>
    </GitSyncTestWrapper>
  )
}
