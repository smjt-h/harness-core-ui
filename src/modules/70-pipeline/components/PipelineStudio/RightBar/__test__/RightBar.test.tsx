/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { act, fireEvent, render, waitFor } from '@testing-library/react'
import { MultiTypeInputType } from '@wings-software/uicore'
import { TestWrapper } from '@common/utils/testUtils'
import { factory } from '@pipeline/components/PipelineSteps/Steps/__tests__/StepTestUtil'
import { Scope } from '@common/interfaces/SecretsInterface'
import { RightBar } from '../RightBar'
import { PipelineContext, PipelineContextInterface } from '../../PipelineContext/PipelineContext'

jest.mock('services/cd-ng', () => ({
  getConnectorPromise: () => Promise.resolve(connectorMock),
  useGetConnector: () => ({
    loading: false,
    data: connectorMock,
    refetch: jest.fn()
  })
}))

jest.mock('../../RightDrawer/RightDrawer', () => ({
  RightDrawer: () => <div />
}))

const connectorMock = {
  status: 'SUCCESS',
  data: {
    connector: {
      name: 'test-connector',
      identifier: 'test_connector',
      description: '',
      orgIdentifier: 'default',
      projectIdentifier: 'projectIdentifier',
      tags: {},
      type: 'Github',
      spec: {
        url: 'https://github.com/',
        validationRepo: 'devrepo',
        authentication: {
          type: 'Http',
          spec: {
            type: 'UsernameToken',
            spec: {
              username: 'username',
              usernameRef: null,
              tokenRef: 'tokenRef'
            }
          }
        },
        type: 'Account'
      }
    }
  }
}

const stateMock = {
  pipeline: {
    name: 'P1',
    identifier: 'P1',
    projectIdentifier: 'Milos2',
    orgIdentifier: 'CV',
    tags: {},
    properties: {
      ci: {
        codebase: {
          connectorRef: 'Git5',
          build: '<+input>'
        }
      }
    },
    stages: [
      {
        stage: {
          name: 'S1',
          identifier: 'S1',
          description: '',
          type: 'CI',
          spec: {
            cloneCodebase: true,
            execution: {
              steps: []
            },
            serviceDependencies: []
          }
        }
      }
    ]
  },
  originalPipeline: {
    name: 'P1',
    identifier: 'P1',
    projectIdentifier: 'Milos2',
    orgIdentifier: 'CV',
    tags: {}
  },
  pipelineIdentifier: '-1',
  pipelineView: {
    isSplitViewOpen: true,
    isDrawerOpened: false,
    splitViewData: {
      type: 'StageView'
    },
    drawerData: {
      type: 'AddCommand'
    }
  },
  gitDetail: {},
  isLoading: false,
  isBEPipelineUpdated: false,
  isDBInitialized: true,
  isUpdated: true,
  isInitialized: true,
  selectionState: {
    selectedStageId: 'S1'
  },
  error: ''
}

export const pipelineContext: PipelineContextInterface = {
  updatePipeline: jest.fn(),
  state: stateMock as any,
  contextType: 'Pipeline',
  allowableTypes: [MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME, MultiTypeInputType.EXPRESSION],
  stepsFactory: factory,
  stagesMap: {},
  isReadonly: false,
  setSchemaErrorView: jest.fn(),
  view: 'ui',
  scope: Scope.PROJECT,
  renderPipelineStage: jest.fn(),
  setView: jest.fn(),
  updateGitDetails: jest.fn(),
  updateEntityValidityDetails: jest.fn(),
  updatePipelineView: jest.fn(),
  updateTemplateView: jest.fn(),
  fetchPipeline: jest.fn(),
  deletePipelineCache: jest.fn(),
  getStageFromPipeline: jest.fn(),
  setYamlHandler: jest.fn(),
  runPipeline: jest.fn(),
  updateStage: jest.fn(),
  pipelineSaved: jest.fn(),
  setSelectedStageId: jest.fn(),
  setSelectedStepId: jest.fn(),
  setSelectedSectionId: jest.fn(),
  setSelection: jest.fn(),
  getStagePathFromPipeline: jest.fn(),
  setTemplateTypes: jest.fn()
}

describe('RightBar', () => {
  test('renders correctly', () => {
    const { container } = render(
      <TestWrapper>
        <PipelineContext.Provider value={pipelineContext}>
          <RightBar />
        </PipelineContext.Provider>
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })

  test('clicking on Variables should open variables view in right drawer', async () => {
    const { getByText } = render(
      <TestWrapper>
        <PipelineContext.Provider value={pipelineContext}>
          <RightBar />
        </PipelineContext.Provider>
      </TestWrapper>
    )
    const variableBtn = getByText('common.variables')
    act(() => {
      fireEvent.click(variableBtn)
    })
    await waitFor(() => expect(pipelineContext.updatePipelineView).toHaveBeenCalled())
    expect(pipelineContext.updatePipelineView).toHaveBeenCalledWith({
      drawerData: {
        type: 'PipelineVariables'
      },
      isDrawerOpened: true,
      isSplitViewOpen: false,
      splitViewData: {}
    })
  })

  test('clicking on Notifications should open notifications table in right drawer', async () => {
    const { getByText } = render(
      <TestWrapper>
        <PipelineContext.Provider value={pipelineContext}>
          <RightBar />
        </PipelineContext.Provider>
      </TestWrapper>
    )
    const notificationsBtn = getByText('notifications.pipelineName')
    act(() => {
      fireEvent.click(notificationsBtn)
    })
    await waitFor(() => expect(pipelineContext.updatePipelineView).toHaveBeenCalled())
    expect(pipelineContext.updatePipelineView).toHaveBeenCalledWith({
      drawerData: {
        type: 'PipelineNotifications'
      },
      isDrawerOpened: true,
      isSplitViewOpen: false,
      splitViewData: {}
    })
  })

  test('clicking on Flow Control should open Flow Control view in right drawer', async () => {
    const { getByText } = render(
      <TestWrapper>
        <PipelineContext.Provider value={pipelineContext}>
          <RightBar />
        </PipelineContext.Provider>
      </TestWrapper>
    )
    const flowControlBtn = getByText('pipeline.barriers.flowControl')
    act(() => {
      fireEvent.click(flowControlBtn)
    })
    await waitFor(() => expect(pipelineContext.updatePipelineView).toHaveBeenCalled())
    expect(pipelineContext.updatePipelineView).toHaveBeenCalledWith({
      drawerData: {
        type: 'FlowControl'
      },
      isDrawerOpened: true,
      isSplitViewOpen: false,
      splitViewData: {}
    })
  })

  test('clicking on Codebase Configuration should open Codebase Configuration view in right drawer', async () => {
    const { getByText } = render(
      <TestWrapper>
        <PipelineContext.Provider value={pipelineContext}>
          <RightBar />
        </PipelineContext.Provider>
      </TestWrapper>
    )
    const codebaseConfigurationBtn = getByText('codebase')
    act(() => {
      fireEvent.click(codebaseConfigurationBtn)
    })
    await waitFor(() => expect(pipelineContext.updatePipelineView).toHaveBeenCalled())
    expect(pipelineContext.updatePipelineView).toHaveBeenCalledWith({
      drawerData: {
        type: 'AddCommand'
      },
      isDrawerOpened: false,
      isSplitViewOpen: false,
      splitViewData: {}
    })
  })

  test('clicking on Advanced Options should open Advanced Options view in right drawer', async () => {
    const { getByText } = render(
      <TestWrapper>
        <PipelineContext.Provider value={pipelineContext}>
          <RightBar />
        </PipelineContext.Provider>
      </TestWrapper>
    )
    const advancedOptionsBtn = getByText('pipeline.advancedOptions')
    act(() => {
      fireEvent.click(advancedOptionsBtn)
    })
    await waitFor(() => expect(pipelineContext.updatePipelineView).toHaveBeenCalled())
    expect(pipelineContext.updatePipelineView).toHaveBeenCalledWith({
      drawerData: {
        type: 'AdvancedOptions'
      },
      isDrawerOpened: true,
      isSplitViewOpen: false,
      splitViewData: {}
    })
  })
})
