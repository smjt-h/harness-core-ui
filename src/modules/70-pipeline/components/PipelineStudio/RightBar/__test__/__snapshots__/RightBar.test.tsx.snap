// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RightBar renders correctly 1`] = `
<div>
  <div
    class="rightBar"
  >
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--withLeftIcon Button--variation Button--variation-tertiary"
      data-testid="input-variable"
      type="button"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
        data-icon="pipeline-variables"
      >
        <svg
          height="28"
          viewBox="0 0 20 20"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M2.197 4.143a.624.624 0 01.564-.356h7.489c.344 0 .624.28.624.624v1.872a.624.624 0 01-1.248 0V5.035H4.072l3.23 3.975c.178.22.186.53.021.76l-3.342 4.626h5.645v-1.248a.624.624 0 111.248 0v1.873c0 .344-.28.624-.624.624H2.76a.624.624 0 01-.505-.99L6.03 9.426 2.276 4.805a.624.624 0 01-.079-.662zM15.9 9.276a5.276 5.276 0 00-.724 1.032c-.22.44-.612.912-.994 1.288-.37.365-.824.732-1.205.884a.624.624 0 11-.463-1.159c.138-.055.445-.272.792-.614.336-.331.62-.692.753-.958.184-.366.529-.88.914-1.308.193-.215.414-.427.646-.59.218-.153.523-.319.87-.319a.624.624 0 01.012 1.248c0 .001-.052.013-.164.092-.129.09-.279.229-.437.404z"
            fill="current-color"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M13.334 9.276c.317.352.596.774.725 1.032.22.44.611.912.993 1.288.37.365.824.732 1.205.884a.624.624 0 10.464-1.159c-.139-.055-.445-.272-.793-.614-.336-.331-.62-.692-.753-.958a6.497 6.497 0 00-.913-1.308 3.98 3.98 0 00-.647-.59c-.217-.153-.523-.319-.87-.319a.624.624 0 00-.012 1.248c0 .001.052.013.165.092.128.09.278.229.436.404z"
            fill="current-color"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        common.variables
      </span>
    </button>
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--withLeftIcon Button--variation Button--variation-tertiary"
      type="button"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
        data-icon="pipeline-deploy"
      >
        <svg
          height="24"
          viewBox="0 0 16 16"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M14.8 1.148a.638.638 0 01.192.577l-2.396 13.707a.614.614 0 01-.43.49.584.584 0 01-.611-.193l-3.813-4.578-1.856 2.382a.584.584 0 01-.666.194.622.622 0 01-.398-.587V8.852l-4.46-2.32a.628.628 0 01-.329-.61.618.618 0 01.423-.543l13.772-4.362a.581.581 0 01.573.131zM6.02 8.783v2.6l4.996-6.413-4.997 3.813zm5.193-5.508L2.24 6.116 5.36 7.74l5.852-4.465zm2.182.62l-4.892 6.278a.59.59 0 01.065.067l3.074 3.69 1.753-10.035z"
            fill="current-color"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        notifications.pipelineName
      </span>
    </button>
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--withLeftIcon Button--variation Button--variation-tertiary"
      type="button"
    >
      <span
        class="bp3-icon bp3-icon-settings StyledProps--main StyledProps--padding-right-xsmall"
        icon="settings"
      >
        <svg
          data-icon="settings"
          height="20"
          viewBox="0 0 20 20"
          width="20"
        >
          <desc>
            settings
          </desc>
          <path
            d="M4 1c0-.55-.45-1-1-1S2 .45 2 1v5h2V1zM2 19c0 .55.45 1 1 1s1-.45 1-1v-6H2v6zm9-18c0-.55-.45-1-1-1S9 .45 9 1v8h2V1zm7 0c0-.55-.45-1-1-1s-1 .45-1 1v3h2V1zM9 19c0 .55.45 1 1 1s1-.45 1-1v-3H9v3zm9-14h-2c-.55 0-1 .45-1 1v3c0 .55.45 1 1 1h2c.55 0 1-.45 1-1V6c0-.55-.45-1-1-1zm-2 14c0 .55.45 1 1 1s1-.45 1-1v-8h-2v8zM4 7H2c-.55 0-1 .45-1 1v3c0 .55.45 1 1 1h2c.55 0 1-.45 1-1V8c0-.55-.45-1-1-1zm7 3H9c-.55 0-1 .45-1 1v3c0 .55.45 1 1 1h2c.55 0 1-.45 1-1v-3c0-.55-.45-1-1-1z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        pipeline.barriers.flowControl
      </span>
    </button>
    <button
      class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--without-shadow Button--withLeftIcon Button--variation Button--variation-tertiary"
      type="button"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
        data-icon="codebase-validating"
      >
        <svg
          height="20"
          viewBox="0 0 24 22"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M23.768 9.589l-5.577-5.9a.81.81 0 10-1.186 1.104l5.066 5.362-5.066 5.362a.808.808 0 00.31 1.311.81.81 0 00.876-.206l5.604-5.928a.808.808 0 00-.027-1.105zM6.954 3.634a.808.808 0 00-1.158.054L.218 9.59a.808.808 0 000 1.105l5.578 5.928a.81.81 0 001.185-1.105l-5.065-5.362L6.98 4.793a.808.808 0 00-.027-1.159zM14.436.02a.74.74 0 00-.569.094.777.777 0 00-.34.477L9.032 19.467a.8.8 0 00.09.584c.105.174.272.3.465.349h.177a.745.745 0 00.46-.158.78.78 0 00.273-.412L14.992.954A.795.795 0 0014.9.37a.76.76 0 00-.465-.35z"
            fill="#6B6D85"
          />
          <rect
            fill="#fff"
            height="11.6"
            rx="5.8"
            stroke="#3E7BFA"
            stroke-width="0.4"
            width="11.6"
            x="12.2"
            y="10.2"
          />
          <path
            clip-rule="evenodd"
            d="M18 13.512a.25.25 0 01-.25-.25V12.25a.25.25 0 11.5 0v1.012a.25.25 0 01-.25.25zm2-.973a.25.25 0 00-.342.091l-.506.877a.25.25 0 10.433.25l.506-.877a.25.25 0 00-.091-.34zm-4.496 2.31a.25.25 0 00.25-.434l-.876-.506a.25.25 0 00-.25.433l.876.506zM16 19.463a.25.25 0 00.341-.092l.507-.876a.25.25 0 00-.433-.25l-.506.876a.25.25 0 00.091.342zm4.155-2.218a.25.25 0 01.341-.092l.877.506a.25.25 0 11-.25.433l-.877-.506a.25.25 0 01-.091-.341zM15.512 16a.25.25 0 01-.25.25H14.25a.25.25 0 110-.5h1.012a.25.25 0 01.25.25zm-.976 2a.25.25 0 00.342.092l.876-.506a.25.25 0 00-.25-.433l-.876.506a.25.25 0 00-.092.342zm5.952-2a.25.25 0 01.25-.25h1.012a.25.25 0 010 .5h-1.012a.25.25 0 01-.25-.25zm-.242-1.583a.25.25 0 10.25.433l.877-.506a.25.25 0 10-.25-.433l-.877.506zM16 12.538a.25.25 0 00-.091.342l.506.877a.25.25 0 10.433-.25l-.506-.877a.25.25 0 00-.342-.09zm2 5.953a.25.25 0 01.25.25v1.012a.25.25 0 01-.5 0V18.74a.25.25 0 01.25-.25zm2 .973a.25.25 0 00.091-.342l-.506-.876a.25.25 0 00-.433.25l.506.876c.07.12.222.16.342.092z"
            fill="#0092E4"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        codebase
      </span>
    </button>
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--withLeftIcon Button--variation Button--variation-tertiary"
      type="button"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
        data-icon="pipeline-advanced"
      >
        <svg
          height="24"
          viewBox="0 0 17 13"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M6.5 1H1v2.375h5.5V1zM1 0a1 1 0 00-1 1v2.375a1 1 0 001 1h5.5a1 1 0 001-1V1a1 1 0 00-1-1H1zm14.25 1h-5.5v2.375h5.5V1zm-5.5-1a1 1 0 00-1 1v2.375a1 1 0 001 1h5.5a1 1 0 001-1V1a1 1 0 00-1-1h-5.5z"
            fill="current-color"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M9.375 3h-2.5V2h2.5v1zm-5.125.917v3.096c0 .416.07.705.172.91.099.198.241.342.435.453.42.24 1.079.325 2.018.325v1c-.936 0-1.84-.072-2.514-.457a2.045 2.045 0 01-.834-.874c-.19-.381-.277-.835-.277-1.357V3.917h1zm9.135 2.417v-.43a.143.143 0 00-.143-.143h-.86a.143.143 0 00-.142.143v.964a.15.15 0 01-.107.14 2.273 2.273 0 00-.388.162.15.15 0 01-.175-.025l-.378-.378-.304-.303a.143.143 0 00-.202 0l-.608.607a.143.143 0 000 .203l.304.304.378.377a.15.15 0 01.024.176 2.283 2.283 0 00-.16.387.15.15 0 01-.142.107h-.964a.143.143 0 00-.143.143v.86c0 .078.064.143.143.143h.964a.15.15 0 01.141.106c.042.135.096.265.161.388a.15.15 0 01-.024.175l-.378.378-.304.304a.143.143 0 000 .202l.608.608a.143.143 0 00.202 0l.304-.304.378-.378a.15.15 0 01.175-.024c.123.065.253.119.388.16a.15.15 0 01.107.142v.964c0 .079.064.143.143.143h.86a.143.143 0 00.142-.143v-.964a.15.15 0 01.107-.141 2.28 2.28 0 00.388-.161.15.15 0 01.175.024l.378.378.304.304a.143.143 0 00.202 0l.608-.608a.143.143 0 000-.202l-.304-.304-.378-.378a.15.15 0 01-.024-.175 2.28 2.28 0 00.16-.388.15.15 0 01.142-.106h.964a.144.144 0 00.143-.144v-.859a.143.143 0 00-.143-.143h-.964a.15.15 0 01-.141-.107 2.283 2.283 0 00-.16-.387.15.15 0 01.023-.176l.378-.377.304-.304a.143.143 0 000-.203l-.608-.607a.143.143 0 00-.202 0l-.304.303-.378.378a.15.15 0 01-.175.025 2.273 2.273 0 00-.388-.161.15.15 0 01-.107-.141v-.534zm-.572 4.01a1.146 1.146 0 100-2.292 1.146 1.146 0 000 2.291zM8.75 9.697H7.5v-1h1.25v1z"
            fill="current-color"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        pipeline.advancedOptions
      </span>
    </button>
    <div />
    <div />
  </div>
</div>
`;
